src/ameba/rule/typing/macro_call_argument_type_restriction.cr

```diff
@@ -0,0 +1,94 @@
+module Ameba::Rule::Typing
+  # A rule that enforces variable arguments to specific macros have a type restriction.
+  #
+  # For example, these are considered invalid:
+  #
+  # ```
+  # class Greeter
+  #   getter name
+  #
+  #   record Task,
+  #     var1 : String,
+  #     var2 = "asdf"
+  # end
+  # ```
+  #
+  # And these are considered valid:
+  #
+  # ```
+  # class Greeter
+  #   getter name : String?
+  #   class_getter age : Int32 = 0
+  #   setter tasks : Array(String) = [] of String
+  #   class_setter queue : Array(Int32)?
+  #   property task_mutex : Mutex = Mutex.new
+  #   class_property asdf : String
+
+  #   record Task,
+  #     var1 : String,
+  #     var2 : String = "asdf"
+  # end
+  # ```
+  #
+  # The `DefaultValue` configuration option controls whether this rule applies to
+  # call arguments that have a default value.
+  #
+  # YAML configuration example:
+  #
+  # ```
+  # Typing/MethodParamTypeRestriction:
+  #   Enabled: false
+  #   DefaultValue: true
+  #   MacroNames:
+  #    - getter
+  #    - getter?
+  #    - getter!
+  #    - class_getter
+  #    - class_getter?
+  #    - class_getter!
+  #    - setter
+  #    - setter?
+  #    - setter!
+  #    - class_setter
+  #    - class_setter?
+  #    - class_setter!
+  #    - property
+  #    - property?
+  #    - property!
+  #    - class_property
+  #    - class_property?
+  #    - class_property!
+  #    - record
+  # ```
+  class MacroCallArgumentTypeRestriction < Base
+    properties do
+      since_version "1.7.0"
+      description "Recommends that variable args to certain macros have type restrictions"
```

Comment:
```suggestion
      description "Recommends that call arguments to certain macros have type restrictions"
```

Reply:
What about `Recommends that variable arguments to certain macros ...`? `call arguments` could be unclear

Reply:
Hmm, what's unclear about it?

Reply:
The `name` in `getter name` and `getter name : String` aren't "calls" (I know the first one technically is, but most people aren't going to know that when they read this).

Reply:
In this case I was referring to `getter` as a _call_, all that follows are the arguments.
To make it more clear, that's the same as writing `getter(foo : String, bar : Int32)`

Reply:
Ahh okay, I thought you were referring to the args as calls, my b.